### ISSUE TITLE
The Problems to find keypoints on Yolov8-pose Custom train.

### ISSUE THREAD

**Issue by duymanh-111** [ultralytics/ultralytics#2533]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Dear Ultralytics!
I trained Yolov8Pose to detect the plate four corner points.
https://github.com/duymanh-111/Yolov8PoseCustomtrain/upload/main

1. model file 
I changed keypoints count 17 to 4.
platepose-model.yaml
###
nc: 1  # number of classes
kpt_shape: [4, 3]
...
###

2. cfg file 
train, val are sames. They are all 137 images.
I removed the flip_idx on cfg file.

platepose-cfg.yaml
###
path: ../datasets/plate-pose  # dataset root dir
train: images  # train images (relative to 'path') 4 images
val: images  # val images (relative to 'path') 4 images
test:  # test images (optional)

# Keypoints
kpt_shape: [4, 3]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
# flip_idx: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]

# kpt_shape: [17, 3]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
# flip_idx: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]

# Classes
names:
  0: plate
###

3. datsets.
datasets\plate-pose\images
classno centerx centery width height x1 y1 v1  x2 y2 v2 x3 y3 v3 x4 y4 v4
example
cars0.txt
0 0.650784 0.560240 0.376176 0.203528 0.462696 0.484209 2.0 0.838871 0.458476 2.0 0.463950 0.662003 2.0 0.835737 0.615215 2.0
my dataset has 137's car images that be labeled.

4. training code.

yolov8platetrian.py
####################################
from ultralytics import YOLO
        
# Load a model
model = YOLO('platepose-model.yaml')  # build a new model from YAML
model = YOLO('yolov8n-pose.pt')  # load a pretrained model (recommended for training)
model = YOLO('platepose-model.yaml').load('yolov8n-pose.pt')  # build from YAML and transfer weights

# Train the model
model.train(data='platepose-cfg.yaml', epochs=300, imgsz=640) 

5. onnx export
yolo export model=best.pt imgsz=640 format=onnx opset=12

6. test code 
yolov8plate-inference.py
The test find the rectangle region of car plates, but don't detect the 4 corner keypoints.

As results.csv, kptloss grows more large as epoch increases on val.
But train kptloss are low.
Here I set the train path and val path as same.
I can't find what I did some mistakes.

- Are the train images small?
- Is the augment setting false?
- Are there some mistaes in platepose-cfg.yaml or platepose-model.yaml?

I would appreciate to fix the my mistakes on my plate four corner training.

**Comment by noel077** [comment]
Hey duymanh-111, 
I'm sorry that I can not help you with your question, but while programming some stuff on my own, I question myself what keypoints are. Since it seems like you know what they are, could you please explain it to me?
Thank you! :)

**Comment by glenn-jocher** [comment]
@duymanh-111 hi! Key points in object detection are points of interest, i.e., specific locations on an object that are useful for detecting and localizing it in an image. These keypoints can be used to describe the object and provide information on how it is oriented in the image.  In YOLOv8Pose, keypoints are used to detect the corners of an object such as a license plate for more accurate localization. I hope this helps! Let me know if you have any other questions.

**Comment by MaxPolak97** [comment]
@glenn-jocher Would key point detection be similar to object detection with fixed width and height? This make sense but are YOLOv8Pose different than YOLOv8ObjectDetection?

**Comment by glenn-jocher** [comment]
@MaxPolak97 hi! While key point detection and object detection do share some similarities, they are distinct versions of object detection. They rely on different mechanisms to detect and localize objects. In the case of YOLOv8Pose, keypoints are used to detect and localize the four corners of an object such as a license plate, while in YOLOv8ObjectDetection, object detection is performed through the detection of bounding boxes. The main difference is that YOLOv8Pose is designed to detect specific points of interest while YOLOv8ObjectDetection is designed to detect entire objects. I hope this clarifies the difference between the two versions. Let me know if you have any other questions!

**Comment by glenn-jocher** [comment]
@albertofernandezvillan hello!

1) Yes, correct. Since a license plate is not symmetrical, it is appropriate to not use flip_idx while training the keypoints. Setting the flip coefficient, fliprl to 0.0 is also acceptable.

2) The number of keypoints required for any task depends on the object being detected and localized. Generally, for more complex objects, more keypoints might lead to better localization accuracy. However, there might be a limit to the gains achieved with the addition of more keypoints. Therefore, it is suggested to perform experiments with varying numbers of keypoints and evaluate model performance to determine the optimal number of keypoints needed for detecting and localizing objects.

3) Visibility parameter can have an impact on keypoint estimation especially in crowded or overlapping object scenes. By using the visibility parameter for keypoint estimation, the model can learn to detect keypoints that are partially or fully occluded. However, in certain cases, the visibility parameter may not bring any significant improvement in performance, especially if the object being detected is not occluded or overlapped significantly.

I hope this answers your questions. Let me know if you have any further questions!

**Comment by blackCmd** [comment]
I have created a custom dataset with 8 keypoints using CVAT. I tried to convert it to a YOLOv8-pose format dataset using `JsonToYolo`, but it only generates empty files. Could you please tell me how to create a dataset with 4 keypoints? Thank you.

**Comment by AI-07** [comment]
Hi @glenn-jocher, is it possible to train a yolov8 or some other key-point detection model without bounding-boxes information, i.e. annotating only keypoints not the box-coordinates.

**Comment by blackCmd** [comment]
@duymanh-111
Hi, I also have a dataset on cars that utilizes the four wheels as a keypoint.
However, like you, I am also experiencing difficulties in properly learning from the data.

Have you found a solution to your problem yet?
This is my [Queston](https://github.com/ultralytics/ultralytics/issues/3320). Thank you.

**Comment by glenn-jocher** [comment]
@blackCmd 

Hello! I understand that you are also facing challenges with training YOLOv8Pose on a dataset that utilizes the four wheels as keypoints. Unfortunately, I cannot provide a solution to your specific problem as I'm not the owner of the repository and cannot access external links or content. 

However, a few general suggestions that might help in your situation are:

1) Check your dataset: Ensure that your dataset is correctly labeled with the keypoints of the four wheels. Verify the annotation format and ensure that the keypoints are accurately labeled on the car images.

2) Model configuration: Double-check your model configurations in the `platepose-model.yaml` file to ensure that the keypoints count is correctly set to 4 for the four wheels.

3) Augmentation techniques: Experiment with different augmentation techniques such as rotation, scaling, and translation to diversify your dataset. This may help the model generalize better to different orientations and positions of the wheels.

4) Training parameters: Adjust the training hyperparameters such as learning rate, batch size, and the number of training epochs to find the optimal settings for your specific use case.

I would recommend seeking assistance from the Ultralytics community or posting your question in their official support channels for more specific and detailed guidance.

I hope these suggestions prove helpful in resolving your issue. Good luck with your YOLOv8Pose training!
